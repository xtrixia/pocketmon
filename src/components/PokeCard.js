import { Link, withRouter } from "react-router-dom";
import { css } from "@emotion/react";
import clsx from "clsx";

import Typography from "./Typography";
import PokeImg from "./PokeImg";

import { COLORS } from "../root/colors";
import { SPACINGS } from "../root/spacings";
import { BREAKPOINTS } from "../root/breakpoints";

const landscape = css``;

const potrait = css`
  flex-direction: column;
  width: 150px;
  @media (min-width: ${BREAKPOINTS.sm}) {
  }
`;

const landscapeImg = css`
  width: 100px;
  @media (min-width: ${BREAKPOINTS.sm}) {
    width: 250px;
  }
`;

const potraitImg = css`
  width: 100%;
`;

const customStyle = css`
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  display: flex;
  border-radius: 5px;
  text-decoration: none;
`;

function PokeCard({
  isClickable,
  actions,
  className,
  description,
  imgUrl,
  title,
  totalOwned,
  type,
  url,
}) {
  let typeStyle = landscape;
  let imgWidth = landscapeImg;
  if (type === "potrait") {
    typeStyle = potrait;
    imgWidth = potraitImg;
  }

  const generateContent = () => {
    return (
      <>
        <PokeImg
          img={imgUrl}
          alt={title}
          type="small"
          className={imgWidth}
          background={COLORS.white2}
        />
        <div
          className={css`
            flex-grow: 1;
            padding: ${SPACINGS.md};
            background: ${COLORS.white};
            display: flex;
            justify-content: center;
            flex-direction: column;
            @media (min-width: ${BREAKPOINTS.sm}) {
              padding: ${SPACINGS.lg};
            }
          `}
        >
          <Typography variant="subheading2">{title}</Typography>
          {description ? (
            generateDescriptionTag(isClickable, description, url)
          ) : (
            <Typography
              variant="body"
              className={css`
                color: ${COLORS.dark};
              `}
            >
              You've owned: {totalOwned}
            </Typography>
          )}
          {actions}
        </div>
      </>
    );
  };

  return isClickable ? (
    <Link to={url} className={clsx(className, typeStyle, customStyle)}>
      {generateContent()}
    </Link>
  ) : (
    <div className={clsx(className, typeStyle, customStyle)}>
      {generateContent()}
    </div>
  );
}

function generateDescriptionTag(isClickable, text, url) {
  return !isClickable ? (
    <Link
      to={url}
      className={css`
        color: ${COLORS.dark};
      `}
    >
      {text}
    </Link>
  ) : (
    <Typography variant="body">{text}</Typography>
  );
}

export default withRouter(PokeCard);
