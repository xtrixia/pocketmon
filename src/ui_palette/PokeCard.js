import { Link, withRouter } from "react-router-dom";
import { css } from "@emotion/css";
import clsx from "clsx";

import Typography from "./Typography";
import PokeImg from "./PokeImg";

import { COLORS } from "../root/colors";
import { SPACINGS } from "../root/spacings";
import { BREAKPOINTS } from "../root/breakpoints";

const landscape = css``;

const potrait = css`
  flex-direction: column;
  width: 125px;
  @media (min-width: ${BREAKPOINTS.sm}) {
  }
`;

const landscapeImg = css`
  width: 100px;
  @media (min-width: ${BREAKPOINTS.sm}) {
    width: 250px;
  }
`;

const potraitImg = css`
  width: 100%;
`;

function PokeCard({
  actions,
  className,
  description,
  imgUrl,
  title,
  totalOwned,
  type,
  url,
}) {
  let typeStyle = landscape;
  let imgWidth = landscapeImg;
  if (type === "potrait") {
    typeStyle = potrait;
    imgWidth = potraitImg;
  }

  return (
    <Link
      to={url}
      className={clsx(
        className,
        typeStyle,
        css`
          filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
          display: flex;
          border-radius: 5px;
          text-decoration: none;
        `
      )}
    >
      <PokeImg
        img={imgUrl}
        alt={title}
        type="small"
        className={imgWidth}
        background={COLORS.white2}
      />
      <div
        className={css`
          flex-grow: 1;
          padding: ${SPACINGS.md};
          background: ${COLORS.white};
          display: flex;
          justify-content: center;
          flex-direction: column;
          @media (min-width: ${BREAKPOINTS.sm}) {
            padding: ${SPACINGS.lg};
          }
        `}
      >
        <Typography variant="subheading2">{title}</Typography>
        {description ? (
          description
        ) : (
          <Typography
            variant="body"
            className={css`
              color: ${COLORS.dark};
            `}
          >
            You've owned: {totalOwned}
          </Typography>
        )}
        {actions}
      </div>
    </Link>
  );
}

export default withRouter(PokeCard);
