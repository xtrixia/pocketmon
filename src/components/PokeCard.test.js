import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";

import PokeCard from "./PokeCard";

test("PokeCard should render landscape and clickable", () => {
  const imgUrl = "https://cdn.logo.com/hotlink-ok/logo-social.png";
  const props = {
    isClickable: true,
    url: `/profile/ivysaur?img=${imgUrl}`,
    title: "ivysaur",
    imgUrl: imgUrl,
    totalOwned: 0,
  };
  const component = renderer.create(
    <Router>
      <PokeCard {...props} />
    </Router>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree.type).toEqual("a");
  expect(tree.props.href).toEqual(`/profile/ivysaur?img=${imgUrl}`);

  const PokeImg = tree.children[0];
  expect(PokeImg.type).toEqual("img");
  expect(PokeImg.props.alt).toEqual(props.title);

  const caption = tree.children[1];
  expect(caption.type).toEqual("div");
  // title
  expect(caption.children[0].children[0]).toEqual(props.title);
  // description
  expect(caption.children[1].children[1]).toEqual(String(props.totalOwned));
});

test("PokeCard should render potrait", () => {
  const imgUrl = "https://cdn.logo.com/hotlink-ok/logo-social.png";
  const props = {
    url: `/profile/ivysaur?img=${imgUrl}`,
    title: "ivysaur",
    imgUrl: imgUrl,
    description: "djoko",
  };
  const component = renderer.create(
    <Router>
      <PokeCard {...props} />
    </Router>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree.type).toEqual("div");

  const PokeImg = tree.children[0];
  expect(PokeImg.type).toEqual("img");
  expect(PokeImg.props.alt).toEqual(props.title);

  const caption = tree.children[1];
  expect(caption.type).toEqual("div");

  // title
  expect(caption.children[0].children[0]).toEqual(props.title);
  // description
  expect(caption.children[1].type).toEqual("a");
  expect(caption.children[1].props.href).toEqual(props.url);
  expect(caption.children[1].children[0]).toEqual(props.description);
});
