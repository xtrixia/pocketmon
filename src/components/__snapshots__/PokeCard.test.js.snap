// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeCard should render landscape and clickable 1`] = `
<a
  className="css-0 css-1cum9qf css-56ur96"
  href="/profile/ivysaur?img=https://cdn.logo.com/hotlink-ok/logo-social.png"
  onClick={[Function]}
>
  <img
    alt="ivysaur"
    className="css-1vwu78b css-vj9s6h css-1uvtpmm"
    src="https://cdn.logo.com/hotlink-ok/logo-social.png"
  />
  <div
    className="css-173zd6x"
  >
    <p
      className="css-ebjej3"
    >
      ivysaur
    </p>
    <p
      className="css-1ohcemi css-qgsv95"
    >
      You've owned: 
      0
    </p>
  </div>
</a>
`;

exports[`PokeCard should render potrait 1`] = `
<div
  className="css-0 css-1cum9qf"
>
  <img
    alt="ivysaur"
    className="css-1vwu78b css-vj9s6h css-1uvtpmm"
    src="https://cdn.logo.com/hotlink-ok/logo-social.png"
  />
  <div
    className="css-173zd6x"
  >
    <p
      className="css-ebjej3"
    >
      ivysaur
    </p>
    <a
      className="css-j662fd css-17iszpw"
      href="/profile/ivysaur?img=https://cdn.logo.com/hotlink-ok/logo-social.png"
      onClick={[Function]}
    >
      djoko
    </a>
  </div>
</div>
`;
